src/workspaces/common/state/useWorkspaceStatePolling.ts

```diff
@@ -0,0 +1,121 @@
+
+// type WorkspaceUpdate = Partial<Workspace['workspace']>
+
+interface WorkspaceUpdate extends Partial<BaseWorkspaceInfo> {
+  workspaceId: string;
+  state: WorkspaceState;
+  errorMessage?: string;
+}
+
+const updateWorkspacesList = (workspaces: Workspace[], update: WorkspaceUpdate): Workspace[] =>
+  workspaces.map((ws) => updateWorkspaceIfMatching(ws, update));
+
+const updateWorkspaceIfMatching = <T extends Workspace>(workspace: T, update: WorkspaceUpdate): T => {
+  if (workspace.workspace?.workspaceId === update.workspaceId) {
+    const updatedWs = _.cloneDeep(workspace);
+    updatedWs.workspace.state = update.state;
+    updatedWs.workspace.errorMessage = update.errorMessage;
+    return updatedWs;
+  }
+  return workspace;
+};
+
+const checkWorkspaceState = async (workspace: Workspace, abort: () => void, signal: AbortSignal) => {
+  const doUpdate = (abort: () => void, update: WorkspaceUpdate) => {
+    workspacesStore.update((wsList) => updateWorkspacesList(wsList, update));
+    workspaceStore.update((ws) => {
+      if (ws) {
+        return updateWorkspaceIfMatching(ws, update);
+      }
+      return undefined;
+    });
+    abort();
+  };
+
+  const startingState = workspace.workspace.state;
+  try {
+    const wsResp: Workspace = await Ajax(signal)
+      .Workspaces.workspace(workspace.workspace.namespace, workspace.workspace.name)
+      .details(['workspace.state', 'workspace.errorMessage']);
+    const state = wsResp.workspace.state;
+
+    if (!!state && state !== startingState) {
+      const update = {
+        workspaceId: workspace.workspace.workspaceId,
+        state,
+        errorMessage: wsResp.workspace.errorMessage,
+      };
+      doUpdate(abort, update);
+    }
+  } catch (error) {
+    // for workspaces that are being deleted, the details endpoint will return a 404 when the deletion is complete
+    if (startingState === 'Deleting' && error instanceof Response && error.status === 404) {
+      const update: WorkspaceUpdate = { workspaceId: workspace.workspace.workspaceId, state: 'Deleted' };
+      doUpdate(abort, update);
+    } else {
+      console.error(`Error checking workspace state for ${workspace.workspace.name}: ${JSON.stringify(error)}`);
+    }
+  }
+};
+
+const updatingStates: WorkspaceState[] = ['Deleting', 'Cloning', 'CloningContainer'];
+
+export const useWorkspaceStatePolling = (workspaces: Workspace[], status: LoadedState<Workspace[]>['status']) => {
+  // we have to do the signal/abort manually instead of with useCancelable so that the it can be cleaned up in the
+  // this component's useEffect, instead of the useEffect in useCancelable
+  const controller = useRef(new window.AbortController());
+  const abort = () => {
+    controller.current.abort();
+    controller.current = new window.AbortController();
+  };
+
+  useEffect(() => {
+    const iterateUpdatingWorkspaces = async () => {
+      if (status === 'Ready') {
```

Comment:
Why do we only update the status if `Ready`? Don't we want to also update when we go from `Cloning` to `CloningContainer` (though I wonder exactly how `Cloning` is related to `CloningContainer`, since it seems like cloning the container is just one step within cloning).

Reply:
Also wondering about `Deleting` -> error state.

Reply:
We don't need to be that granular for users. We just need to know if it's cloning (any cloning step), ready, or error.

Reply:
Perhaps the UI doesn't need it now, but I would expect the state to be correct on the stored workspaces, and obviously we need to handle error. But maybe I'm not understanding what this code is doing.

Reply:
Oh wow, yeah, that's super confusing. Perhaps naming the variable something else would help-- like dataLoadedStatus?